doc/developer-guide/io-framework.md

```diff
@@ -0,0 +1,294 @@
+   flushed.
+
+### Available I/O operations
+
+The I/O framework supports two ways of sending operations to the kernel.
+In direct mode, each request is sent independently of the others. In batch
+mode multiple requests are sent together all at once.
+
+All operations will also have a `data` argument to pass any additional
+per-request private data that the callback may need. This data will be
+available in `op->data` for most of the cases (there's an exception for
+asynchronous requests. See later).
+
+Many of the I/O operations will have a timeout argument, which represents
+the maximum time allowed for the I/O to complete. If the operation takes
+more than that time, the system call will be cancelled and the callback
+will be executed passing a `-ETIMEDOUT` error.
+
+I/O operations will also have a priority argument that makes it possible
+to give different priorities to each requests so that the kernel scheduler
+can efficiently manage them based on their priority.
+
+An identifier is returned for each request. This value can be used to try
+to cancel the associated request if it has not been started or completed
+yet.
+
+#### Direct mode
+
+In direct mode the interface is really simple. Each function only requires
+the data needed to perform the operation and returns an identifier. No
+memory allocations are needed.
+
+#### Batch mode
+
+In batch mode a `gf_io_batch_t` object needs to be created, which will
+contain all requests to send. Then one or more `gf_io_request_t` objects
+need to be created and added to the batch object. Both types of objects
+can be allocated in the stack because they are not needed once the batch
+is submitted.
+
+The functions to prepare requests have the same name as those used in
+direct mode but with the `_prepare` suffix. The function signature is
+exactly the same, but adding a `gf_io_request_t` argument.
+
+Once a request is prepared, it can be added to the batch object using
+`gf_io_batch_add()`. This function also receives a pointer to an identifier.
+If it's not NULL, the id of this request will be copied to that location
+once the batch is submitted.
+
+Optionally, it's possible to create a chain of dependencies between requests
+of a batch. In this case, a chained request will only be executed once the
+previous request has finished with a success.
+
+Once the batch is ready, it can be processed by calling `gf_io_batch_submit()`.
+
+##### Operations
+
+###### Cancel request
+
+```c
+uint64_t
+gf_io_cancel(gf_io_callback_t cbk, uint64_t ref, void *data);
+```
+
+Tries to cancel the request identified by `ref`. `cbk` will be called with
+error 0 if the request has been cancelled, `-ENOENT` if the request cannot
+be found (it has already terminated probably), or `-EALREADY` if the request
+is still there but cannot be cancelled.
+
+In case that the request can be successfully cancelled, the callback
+associated to that request will be called with error `-ECANCELED`.
```

Comment:
When you want to cancel a previous request, there are two involved requests:

1. The original request we want to cancel.
2. The request to cancel the previous one.

If the original request has already been started or completed, the second request will fail with some error because it has failed to cancel it. However if the first request is still pending, the second request will succeed (it will return 0). In this case, the first request will also terminate, but with error `-ENCACELED` to indicate that someone cancelled it.